descriptors: MuSig2

achow101 opened the pull request:
Implements parsing of BIP 390 `musig()` descriptors.

Split from #29675

achow101 committed:
script/parsing: Allow Const to not skip the found constant

When parsing a descriptor, it is useful to be able to check whether a
string begins with a substring without consuming that substring as
another function such as Func() will be used later which requires that
substring to be present at the beginning.

Specifically, for MuSig2, this modified Const will be used to determine
whether a an expression begins with "musig(" before a subsequent
Func("musig", ...) is used.

achow101 committed:
descriptors: Add PubkeyProvider::IsBIP32()

achow101 committed:
descriptor: Add MuSigPubkeyProvider

achow101 committed:
descriptor: Parse musig() key expressions

achow101 committed:
tests: Test musig() parsing

achow101 committed:
Add MuSig2 Keyagg Cache helper functions

secp256k1 provides us secp256k1_musig_keyagg_cache objects which we are
used as part of session info and to get the aggregate pubkey. These
helper functions help us convert to/from the secp256k1 C objects into
the Bitcoin Core C++ objects.

achow101 committed:
sign: Add GetMuSig2ParticipantPubkeys to SigningProvider

achow101 committed:
XOnlyPubKey: Add GetCPubKeys

We need to retrieve the even and odd compressed pubkeys for xonly
pubkeys, so add a function to do that. Also reuse it in GetKeyIDs.

achow101 committed:
build: Enable secp256k1 musig module

dergoegge commented:
```
$ echo "cGsobXVzaWcoZGR9dXVlLzAwLylrKA==" | base64 --decode > mocked_descriptor_parse_1.crash
$ FUZZ=mocked_descriptor_parse fuzz mocked_descriptor_parse_1.crash
script/descriptor.cpp:1838 ParsePubkey: Assertion `Func("musig", expr)' failed.

$ echo "dHIobXVzaWcoICAgICB0dXVzKG9sZGVwayhnZylnZ2dnZmdnKTwseigoKCgoKCgoKCgoKCgoKCgoKCgoKHN0KQ==" | base64 --decode > mocked_descriptor_parse_2.crash
$ FUZZ=mocked_descriptor_parse fuzz mocked_descriptor_parse_2.crash
script/descriptor.cpp:1833 ParsePubkey: Assertion `split.size() <= 2' failed.
```

achow101 commented:
Fixed the fuzz crashes and added those descriptors to the unit tests.

dergoegge commented:
```
$ echo "dHIobXVzaWcoJTIyLzMzMmgpSigoKChhZGRyKEJjdXUp" | base64 --decode > mocked_descriptor_parse_3.crash
$ FUZZ=mocked_descriptor_parse fuzz mocked_descriptor_parse_3.crash
script/descriptor.cpp:632 GetPubKey: Assertion `pubkey.has_value()' failed.
```

dergoegge commented:
```
$ echo "dHIobXVzaWcoKS8wMDAwMTEp" | base64 --decode > descriptor_parse_1.crash
$ FUZZ=descriptor_parse fuzz descriptor_parse_1.crash
[libsecp256k1] illegal argument: pubkeys != NULL
```

dergoegge commented:
```
$ echo "dHIobXVzaWcoJWU5LzwwOzAwMDAwMzU7MDYwOzM7MDY+KS82Nik=" | base64 --decode > mocked_descriptor_parse_4.crash
$ FUZZ=mocked_descriptor_parse fuzz mocked_descriptor_parse_4.crash
fuzz: script/descriptor.cpp:1942: std::vector<std::unique_ptr<PubkeyProvider>> (anonymous namespace)::ParsePubkey(uint32_t &, const Span<const char> &, ParseScriptContext, FlatSigningProvider &, std::string &): Assertion `pub.size() == 1' failed.
```

achow101 committed:
util/string: Allow Split to include the separator

When splitting a string, sometimes the separator needs to be included.
Split will now optionally include the separator at the end of the left
side of the splits, i.e. it appears at the end of the splits, except
for the last one.

Specifically, for musig() descriptors, Split is used to separate a
musig() from any derivation path that follows it by splitting on the
closing parentheses. Since that parentheses is needed for Func() and
Expr(), Split() needs to preserve the end parentheses instead of
discarding it.

achow101 commented:
All dependencies have been merged, ready for review.

theStack commented on src/script/signingprovider.cpp:
```diff
@@ -82,13 +87,21 @@ bool FlatSigningProvider::GetTaprootBuilder(const XOnlyPubKey& output_key, Tapro
 
+std::vector<CPubKey> FlatSigningProvider::GetAggregateParticipantPubkeys(const CPubKey& pubkey) const
+{
+    const auto& it = aggregate_pubkeys.find(pubkey);
+    if (it == aggregate_pubkeys.end()) return {};
+    return it->second;
```
nit: could use `LookupHelper` here (for slightly better readability imho)
```suggestion
    std::vector<CPubKey> participant_pubkeys;
    LookupHelper(aggregate_pubkeys, pubkey, participant_pubkeys);
    return participant_pubkeys;
```

achow101 replied:
Done

theStack commented on src/musig.h:
```diff
@@ -0,0 +1,18 @@
+
+#include <pubkey.h>
+
+#include <vector>
```
nit: should also include ` <optional>`, since it's used for the return types below
```suggestion
#include <optional>
#include <vector>
```

achow101 replied:
Done

theStack commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,219 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+    bool IsRangedDerivation() const { return m_derive != DeriveType::NO; }
+    bool IsRangedParticipants() const
+    {
+        for (const auto& pubkey : m_participants) {
+            if (pubkey->IsRange()) return true;
+        }
+        return false;
```
nit: could turn this into a one-liner by using `std::any_of`

achow101 replied:
Done

theStack commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,219 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+    bool IsBIP32() const override
+    {
+        // musig() can only be a BIP 32 key if all participants are bip32 too
+        bool all_bip32 = true;
+        for (const auto& pk : m_participants) {
+            all_bip32 &= pk->IsBIP32();
+        }
+        return all_bip32;
```
nit: could turn this into a one-liner by using `std::all_of`

achow101 replied:
Done

theStack reviewed (commented):
Concept ACK

Seems like the commit message for 2e6dcdbc8055660a2e20ba81b62b7d26ae0ccb05 ("Add MuSig2 Keyagg Cache class and functions") is out-of-sync, as there is no such class added and also the mentioned `MuSig2KeyAggCacheImpl` doesn't exist.

For the newly introduced parameters of the `Const` and `Split` functions (commits fe02d7cb237c00de6abe1776b3101342ffddf757 and 4a1eeee27a64c4be7293740f8fff839879b88d86), it would be nice to have unit test coverage, but this can be also done in a follow-up.

w0xlt reviewed (commented):
Concept ACK.

achow101 commented:
> Seems like the commit message for [2e6dcdb](https://github.com/bitcoin/bitcoin/commit/2e6dcdbc8055660a2e20ba81b62b7d26ae0ccb05) ("Add MuSig2 Keyagg Cache class and functions") is out-of-sync, as there is no such class added and also the mentioned `MuSig2KeyAggCacheImpl` doesn't exist.

Updated the commit message.

theStack commented on src/musig.h:
```diff
@@ -0,0 +1,19 @@
+
+struct secp256k1_musig_keyagg_cache;
+
+bool GetMuSig2KeyAggCache(const std::vector<CPubKey>& pubkeys, secp256k1_musig_keyagg_cache& keyagg_cache);
+std::optional<CPubKey> GetCPubKeyFromMuSig2KeyAggCache(secp256k1_musig_keyagg_cache& cache);
```
note for other reviewers that might wonder: these two functions and the involved `secp256k1_musig_keyagg_cache` instance are not publicly used within this PR yet, but they will be in #29675 (or any other future PRs that get split up from that one)

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,212 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            prov->GetPrivKey(pos, arg, out);
+        }
+    }
+    std::optional<CPubKey> GetRootPubKey() const override
```
In 6b634754d2783ffe16570ad37dbcf9251a7efc24 "descriptor: Add MuSigPubkeyProvider": could add a short comment to explain why this doesn't exist

achow101 replied:
Done

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,212 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+
+    void GetPrivKey(int pos, const SigningProvider& arg, FlatSigningProvider& out) const override
+    {
+        // Get the private keys for all participants
```
In 6b634754d2783ffe16570ad37dbcf9251a7efc24 "descriptor: Add MuSigPubkeyProvider": do you mean "all or nothing" or "any that we have"?

achow101 replied:
any that we have.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,212 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            }
+        }
+        out += ")";
+        out += FormatHDKeypath(m_path, /*apostrophe=*/true);
```
In https://github.com/bitcoin/bitcoin/commit/6b634754d2783ffe16570ad37dbcf9251a7efc24 "descriptor: why apostrophe? 😢

Sjors replied:
In that case it's less confusing if you don't pass this argument (as the default is false).

Sjors replied:
Update: I see you already did that in 048c95ea55.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,212 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+                extpub.nDepth = 0;
+                std::memset(extpub.vchFingerprint, 0, 4);
+                extpub.nChild = 0;
+                extpub.chaincode.FromHex("6589e367712c6200e367717145cb322d76576bc3248959c474f9a602ca878086");
```
This constant should probably live in `musig.h`.

achow101 replied:
Done, added to musig.h

Sjors reviewed (commented):
Concept ACK

Reviewed up to 6b634754d2783ffe16570ad37dbcf9251a7efc24, mostly happy.

I didn't thoroughly check that all test vectors from BIP390 are in the test, but did check a few.

> Can you add an example to `doc/descriptors.md`?

This still seems useful (and this document should probably link back to the bip numbers, but that's orthogonal).

Manual testing hint for other reviewers: you can use the `deriveaddresses` RPC with some of the test vectors. Just use `00000000` as the checksum, it will tell you the right one, and then call it gain to see an address. Pass that address into `validateaddress` and compare the `scriptPubKey` to the one in the bip.

achow101 commented on src/script/parsing.h:
```diff
@@ -13,10 +13,10 @@ namespace script {
 /** Parse a constant.
  *
- * If sp's initial part matches str, sp is updated to skip that part, and true is returned.
+ * If sp's initial part matches str, sp is optionally updated to skip that part, and true is returned.
```
**sp**an?

achow101 commented on src/musig.h:
```diff
@@ -0,0 +1,19 @@
+
+bool GetMuSig2KeyAggCache(const std::vector<CPubKey>& pubkeys, secp256k1_musig_keyagg_cache& keyagg_cache);
+std::optional<CPubKey> GetCPubKeyFromMuSig2KeyAggCache(secp256k1_musig_keyagg_cache& cache);
+std::optional<CPubKey> MuSig2AggregatePubkeys(const std::vector<CPubKey>& pubkeys);
```
"Aggregate" is used as an imperative verb here, so it is "(you,) aggregate (these) pubkeys"

Sjors replied:
Ah, it's a verb, then it makes sense.

achow101 committed:
doc: Add musig() example

achow101 commented on src/pubkey.h:
```diff
@@ -286,6 +286,7 @@ class XOnlyPubKey
      * This is needed for key lookups since keys are indexed by CKeyID.
      */
     std::vector<CKeyID> GetKeyIDs() const;
+    std::vector<CPubKey> GetCPubKeys() const;
```
Updated the comments.

achow101 commented on src/util/string.h:
```diff
@@ -100,18 +100,27 @@ void ReplaceAll(std::string& in_out, const std::string& search, const std::strin
  *
  * If sep does not occur in sp, a singleton with the entirety of sp is returned.
  *
+ * @param[in] include_sep Whether to include the separator at the end of the left side of the splits.
```
Done

achow101 commented on src/util/string.h:
```diff
@@ -100,18 +100,27 @@ void ReplaceAll(std::string& in_out, const std::string& search, const std::strin
  * "foo(bar(1),2),3) on ',' will return {"foo(bar(1)", "2)", "3)"}.
+ *
+ * If include_sep == true, splitting "foo(bar(1),2),3) on ','
+ * will return {"foo(bar(1),", "2),", "3)"}.
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -221,6 +221,9 @@ struct PubkeyProvider
     /** Make a deep copy of this PubkeyProvider */
     virtual std::unique_ptr<PubkeyProvider> Clone() const = 0;
+
+    /** Whether this PubkeyProvider can be a BIP 32 extended key that can be derived from */
```
"is"

achow101 commented on src/musig.h:
```diff
@@ -0,0 +1,19 @@
+
+struct secp256k1_musig_keyagg_cache;
+
+bool GetMuSig2KeyAggCache(const std::vector<CPubKey>& pubkeys, secp256k1_musig_keyagg_cache& keyagg_cache);
```
Added comments.

theStack commented on src/musig.h:
```diff
@@ -12,6 +12,10 @@
 struct secp256k1_musig_keyagg_cache;
 
+//! MuSig2 chaincode as defined by BIP 328
+//! uint256 will byteswap the hex
+constexpr uint256 MUSIG_CHAINCODE{"6589e367712c6200e367717145cb322d76576bc3248959c474f9a602ca878086"};
```
in commit 048c95ea557cf81a8e55c5ffe9994cc9d8f3e039: not too fond of "reverse-hexstrings of byte-strings"  creeping in, could do instead
```suggestion
using namespace util::hex_literals;
constexpr std::array<uint8_t, 32> MUSIG_CHAINCODE{"868087ca02a6f974c4598924c36b57762d32cb45717167e300622c7167e38965"_hex_u8};
```
and assign the constant to the uint256 instance via `... = uint256(MUSIG_CHAINCODE)`.
(but shouldn't be a blocker, can be improved in a follow-up as well, if others are even as annoyed by that as me 😅 )

Sjors replied:
Indeed it's nice if we can avoid this, because although reviewing the reversal was easy, finding it later with search is tedious.

achow101 replied:
It looks like `_hex_u8` can be used directly in uin256's constructor with normal byte order, so used that.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,219 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            if (!pub) return std::nullopt;
+            pubout = *pub;
+            out.aggregate_pubkeys.emplace(m_aggregate_pubkey.value(), pubkeys);
+        } else if (IsRangedParticipants()) {
```
Done, with `Assume`. It's occurred to me that `Assert` in descriptor code is probably not the way to go, so I've changed the couple other `Assert` to `Assume`.

achow101 commented on src/musig.h:
```diff
@@ -0,0 +1,27 @@
+using namespace util::hex_literals;
+constexpr uint256 MUSIG_CHAINCODE{"868087ca02a6f974c4598924c36b57762d32cb45717167e300622c7167e38965"_hex_u8};
+
+//! Create a secp56k1_musig_keyagg_cache from the pubkeys in their current order. This is necessary for most MuSig2 operations
```
Done

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,220 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+
+    std::optional<CPubKey> GetPubKey(int pos, const SigningProvider& arg, FlatSigningProvider& out, const DescriptorCache* read_cache = nullptr, DescriptorCache* write_cache = nullptr) const override
+    {
+        // If the participants are not ranged, we can compute and cache the aggregate pubkey by creating a PubkeyProvider for it
```
In 35db4f2dcfc3435e10935581ffa447ffe219cc1e "descriptor: Add MuSigPubkeyProvider": just to check my own understanding, by ranged participants you mean the case of `musig(alice/*, bob)`, i.e. derivation before aggregation?

I assume that if the participants are ranged, then you can't also have a range after derivation, e.g. `musig(alice/*, bob)/*` isn't allowed. Where do we check this?

achow101 replied:
In the next commit which implements parsing.

Sjors replied:
Being able to generate a (derived) descriptor without parsing is still on my wish list: https://github.com/bitcoin/bitcoin/issues/24003

But even in that case it could generate, convert to string and then re-parse as a sanity check.

achow101 replied:
Added an `Assume`.

Sjors reviewed (commented):
Some more questions about the `MuSigPubkeyProvider` introduced in 35db4f2dcfc3435e10935581ffa447ffe219cc1e.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,220 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        std::vector<std::unique_ptr<PubkeyProvider>> providers;
+        providers.reserve(m_participants.size());
+        for (const std::unique_ptr<PubkeyProvider>& p : m_participants) {
+            providers.emplace_back(p->Clone());
```
It is used for multipath descriptors, tests of which are added 2 commits later.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,220 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        bool any_privkeys = false;
+        out = "musig(";
+        for (size_t i = 0; i < m_participants.size(); ++i) {
+            const auto& pubkey = m_participants.at(i);
```
I find that harder to read. `for i` is used here because we need the position.

theStack commented on src/script/descriptor.cpp:
```diff
@@ -1802,9 +1807,153 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
 /** Parse a public key including origin information (if enabled). */
-std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkey(uint32_t key_exp_index, const std::span<const char>& sp, ParseScriptContext ctx, FlatSigningProvider& out, std::string& error)
+// NOLINTNEXTLINE(misc-no-recursion)
+std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkey(uint32_t& key_exp_index, const std::span<const char>& sp, ParseScriptContext ctx, FlatSigningProvider& out, std::string& error)
```
Why is the `key_exp_index` parameter changed to pass by reference? Without that the unit tests fail, so it seems necessary, but I'm confused why this was not needed earlier already for similar constructs (e.g. for `multi()` expressions where the value is also incremented in the key expression parsing loop). Probably I'm missing some basic descriptors parsing knowledge.

theStack reviewed (commented):
Reviewed thoroughly up to ff7476ec32f4dbee07bf04169e741e2aea5a9ab7, looks good to me (with one refactoring suggestion below). Still need to go again through the parsing code as there is a lot going on with all the conditions on what is allowed and what not w.r.t. derivations.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,224 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+                    tmp = pubkey->ToString(PubkeyProvider::StringType::COMPAT);
+                    break;
+            }
+            out += tmp;
```
Good point, done.

theStack reviewed (commented):
According to BIP 390 [_"Repeated participant public keys are not allowed."_](https://github.com/bitcoin/bips/blob/72af87fc72999e3f0a26a06e6e0a7f3134236337/bip-0390.mediawiki?plain=1#L36), but it seems there is currently no check preventing that? Example test case that passes but shouldn't (IIUC): https://github.com/theStack/bitcoin/commit/f260c7ec06e8bbb7148877a9a9b7e96707c41fa1

achow101 commented:
> According to BIP 390 [_"Repeated participant public keys are not allowed."_](https://github.com/bitcoin/bips/blob/72af87fc72999e3f0a26a06e6e0a7f3134236337/bip-0390.mediawiki?plain=1#L36), but it seems there is currently no check preventing that? Example test case that passes but shouldn't (IIUC): [theStack@f260c7e](https://github.com/theStack/bitcoin/commit/f260c7ec06e8bbb7148877a9a9b7e96707c41fa1)

Good catch. Fixed and added a test.

theStack commented on src/script/descriptor.cpp:
```diff
@@ -1797,9 +1803,159 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+            error = "Too many ')' in musig() expression";
+            return {};
+        }
+        std::span<const char> sp_musig(split.at(0).begin(), split.at(0).end());
+
+        auto expr = Expr(sp_musig);
+        if (!Func("musig", expr)) {
```
in commit 9e3a1d05bf066df31a6e124638046e70669eb470: nit: the `Expr` call seems to be a no-op here, considering that we already extracted the function expression with the parens-splitting above
```suggestion
        std::span<const char> expr(split.at(0).begin(), split.at(0).end());
        if (!Func("musig", expr)) {
```

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,215 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+    const DeriveType m_derive;
+
+    bool IsRangedDerivation() const { return m_derive != DeriveType::NO; }
+    bool IsRangedParticipants() const
+    {
+        return std::any_of(m_participants.begin(), m_participants.end(), [](const auto& pubkey) { return pubkey->IsRange(); });
+    }
```
`IsRangedParticipants` is called multiple times in this class, can evaluate it once in the constructor itself and store the result in a const. I don't suppose a `MuSigPubkeyProvider` is updated with more `m_participants` after creation.

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,215 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        CPubKey pubout;
+        if (m_aggregate_provider) {
+            // When we have a cached aggregate key, we are either returning it or deriving from it
+            // Either way, we can passthrough to it's GetPubKey
```
```diff
- Either way, we can passthrough to it's GetPubKey
+ Either way, we can passthrough to its GetPubKey
```

rkrux reviewed (commented):
Partial code review 981b2abd79e3a7a4701c9d8794057b5523acb177
I just started and am still reviewing.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1797,9 +1803,159 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+                error = strprintf("musig(): %s", error);
+                return {};
+            }
+            const auto& [_, inserted] = key_exprs.emplace(arg.begin(), arg.end());
+            if (!inserted) {
+                error = strprintf("musig(): Cannot have repeated participant key expressions");
+                return {};
+            }
```
> allow duplicate keys or more thorowly prevent them.

Yes, I've suggest to the [mailing list](https://groups.google.com/g/bitcoindev/c/SSpyvbD9CMg) that we should allow the duplicate keys. Still checking with cryptographers that that will be okay, particularly since PSBT doesn't allow duplicate participants to use different pubnonces and partial sigs.

achow101 replied:
> Good point, these 2 structs in PSBT would need to be updated: 

I believe there won't need to be any changes to PSBT. If a participant pubkey is duplicated, it is still valid to for each time the pubkey appears, to use the same pubnonce and the same partial sig for that nonce for each instance the pubkey appears. The only complication with this is if the duplicate pubkeys are different signers and produce different nonces and partial sigs as they may overwrite each other resulting a PSBT with a partial sig that does not match the pubn However, I'm pretty sure (and I did check with a cryptographer) that there cannot be any key leakage.

rkrux replied:
I see that I had misunderstood your previous comment, this^ looks fine to me.

> The Sign algorithm must not be executed twice with the same secnonce. Otherwise, it is possible to extract the secret signing key from the two partial signatures output by the two executions of Sign.

This^ is the only precaution I read in BIP 327 related to reusing, and I don't think anything in the above comment goes against this precaution. Specially since using the same partial sig doesn't translate to signing again.

I have not yet thought through the consequences when PSBTs are shared between signers during the signing process, I believe that picture would be clearer to me in the next PR of Musig signing.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1797,9 +1803,157 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+        };
+
+        if (max_providers_len > 1 && paths.size() > 1) {
+            error = "musig(): Cannot have multipath participant keys if musig() is also multipath";
```
Since multipath is similar to ranged expansion, I've applied the same restrictions that are on ranged to multipath as well. I suppose this should be mentioned in the BIP.

rkrux replied:
I raised a PR for it here: https://github.com/bitcoin/bips/pull/1866

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1797,9 +1803,157 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+        std::vector<KeyPath> paths;
+        if (split.size() == 2 && Const("/", split.at(1), /*skip=*/false)) {
+            if (!all_bip32) {
+                error = "musig(): musig() derivation requires all participants to be xpubs";
```
Done

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,215 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            } else {
+                m_aggregate_provider = std::make_unique<ConstPubkeyProvider>(m_expr_index, m_aggregate_pubkey.value(), /*xonly=*/false);
+            }
+        }
```
```cpp
prov->GetPubKey(0, arg,
```

Ah I see now the SigningProvider being passed here. I notice that `const SigningProvider& arg` is quite prevalent in this file, but I don't prefer calling the signing provider here just `arg` - it's easy to miss. Fine to stick with it in this PR, can be updated separately for all occurrences in one go if it makes sense to others as well.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -574,6 +581,216 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            }
+        }
+
+        // Retrieve all participant pubkeys
+        std::vector<CPubKey> pubkeys;
+        for (const auto& prov : m_participants) {
+            std::optional<CPubKey> pub = prov->GetPubKey(pos, arg, out, read_cache, write_cache);
+            if (!pub) return std::nullopt;
+            pubkeys.emplace_back(*pub);
+        }
+        std::sort(pubkeys.begin(), pubkeys.end());
```
Even with aggregate derivation, `out` needs to be filled with participant pukey info. This loop is present in order to do that for both aggregate derivation and participant derivation.

achow101 replied:
Non-ranged participants generally won't be rederived every time because of the `read_cache`.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -574,6 +581,216 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        if (m_aggregate_provider) {
+            // When we have a cached aggregate key, we are either returning it or deriving from it
+            // Either way, we can passthrough to its GetPubKey
+            std::optional<CPubKey> pub = m_aggregate_provider->GetPubKey(pos, arg, out, read_cache, write_cache);
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -574,6 +581,216 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            throw std::runtime_error("musig(): Cannot have both ranged participants and ranged derivation");
+        }
+        if (!Assume(m_derive != DeriveType::HARDENED)) {
+            throw std::runtime_error("musig(): Cannot have hardened hardened derivation");
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -574,6 +581,216 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            out.aggregate_pubkeys.emplace(m_aggregate_pubkey.value(), pubkeys);
+        } else {
+            if (!Assume(m_ranged_participants)) return std::nullopt;
+            // Derive participants and compute new aggregate key
```
Changed the comment.

w0xlt commented on src/script/parsing.h:
```diff
@@ -13,10 +13,10 @@ namespace script {
  * Otherwise sp is unmodified and false is returned.
  */
-bool Const(const std::string& str, std::span<const char>& sp);
+bool Const(const std::string& str, std::span<const char>& sp, bool skip = true);
```
Maybe a test in `src/test/util_tests.cpp` in `BOOST_AUTO_TEST_CASE(test_script_parsing)` would be good for consistency's sake.

achow101 replied:
Done

w0xlt reviewed (commented):
Perhaps the description of the commits `script/parsing: Allow Const to not skip the found constant ` (https://github.com/bitcoin/bitcoin/pull/31244/commits/4b135f80d03c92fb522d1dcfd7d697aa0a4af626) and  `util/string: Allow Split to include the separator` (https://github.com/bitcoin/bitcoin/pull/31244/commits/b89a937225ed217556b38e000ee5d1bf0b5952aa) could explain why these changes are necessary for the musig2 descriptor.

w0xlt commented on src/script/signingprovider.h:
```diff
@@ -161,6 +161,7 @@ class SigningProvider
     virtual bool GetKeyOrigin(const CKeyID& keyid, KeyOriginInfo& info) const { return false; }
     virtual bool GetTaprootSpendData(const XOnlyPubKey& output_key, TaprootSpendData& spenddata) const { return false; }
     virtual bool GetTaprootBuilder(const XOnlyPubKey& output_key, TaprootBuilder& builder) const { return false; }
+    virtual std::vector<CPubKey> GetAggregateParticipantPubkeys(const CPubKey& pubkey) const { return {}; }
```
Without proper context, it may not be clear that this aggregation specifically refers to a MuSig2 aggregation.
Perhaps a name change could make the purpose and use of the function clearer.
```suggestion
    virtual std::vector<CPubKey> GetMuSig2AggregateParticipantPubkeys(const CPubKey& pubkey) const { return {}; }
```

achow101 replied:
Changed to `GetMuSig2ParticipantPubkeys`

achow101 commented on src/musig.cpp:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    // Aggregate the pubkey
+    if (!secp256k1_musig_pubkey_agg(secp256k1_context_static, nullptr, &keyagg_cache, pubkey_ptrs.data(), pubkey_ptrs.size())) {
```
That can be added if it is needed.

All of the MuSig2 operations rely on the keyagg cache, which is only produced by aggregating. That's why this function has that as an output parameter.

achow101 commented:
> Perhaps the description of the commits `script/parsing: Allow Const to not skip the found constant ` ([4b135f8](https://github.com/bitcoin/bitcoin/commit/4b135f80d03c92fb522d1dcfd7d697aa0a4af626)) and `util/string: Allow Split to include the separator` ([b89a937](https://github.com/bitcoin/bitcoin/commit/b89a937225ed217556b38e000ee5d1bf0b5952aa)) could explain why these changes are necessary for the musig2 descriptor.

Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -574,6 +581,218 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+                }
+                pubkeys.push_back(pubkey.value());
+            }
+            std::sort(pubkeys.begin(), pubkeys.end());
```
Yes. They both essentially do memcmp on the serialized pubkey

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/31244/commits/f1b0fded41223c9b5b92696a0cfa553a10f3d9bf

theStack reviewed (approved):
Code-review ACK f1b0fded41223c9b5b92696a0cfa553a10f3d9bf
(modulo the outstanding question on how to cope with duplicate participant pubkeys, as https://github.com/bitcoin/bips/pull/1867 isn't merged yet; very likely the repeated key expression detection in 1e738242a8c642805b7cdbab51a21ee5a298bbab can be removed)

achow101 commented:
#29675 revealed a bug in `ToNormalizedString` which is fixed in the latest push.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/31244/commits/1e205ca66c97295eca21f61dc309e8403814e802

theStack reviewed (approved):
re-ACK 1e205ca66c97295eca21f61dc309e8403814e802

achow101 commented:
The BIP change was merged, so dropped the duplicate check

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1800,9 +1806,157 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+                return {};
+            }
+            bool dummy = false;
+            if (!ParseKeyPath(deriv_split, paths, dummy, error, /*allow_multipath=*/true, /*allow_hardened=*/false)) {
```
Done as suggested.

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/31244/commits/f3c97b81a214724c14d8c8a7db1207f19584db83

theStack commented on src/script/descriptor.cpp:
```diff
@@ -1799,10 +1800,168 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+                return {};
+            }
+            for (size_t i = 0; i < max_providers_len; ++i) {
+                // Final MuSigPubkeyProvider use participant pubkey providers at each multipath position, and the first (and only) path
```
```suggestion
                // Final MuSigPubkeyProvider uses participant pubkey providers at each multipath position, and the first (and only) path
```

achow101 replied:
Done

theStack commented on src/script/descriptor.cpp:
```diff
@@ -1799,10 +1800,168 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+                emplace_final_provider(i, i);
+            }
+        } else {
+            // No multipath derivation MuSigPubkeyProvider uses the first (and only) participant pubkey providers, and the first (and only) path
```
```suggestion
            // No multipath derivation, MuSigPubkeyProvider uses the first (and only) participant pubkey providers, and the first (and only) path
```

achow101 replied:
Done

theStack reviewed (approved):
re-ACK 35a17e5a2e1394cc3f4f5a4f304f58528cb44151 :musical_note: :two: 

Verified that since my previous ACK, the duplicate KEY expression check has been removed (as per BIP390 change https://github.com/bitcoin/bips/pull/1867), test cases were updated accordingly, and the [alternative "disallow hardened derivation" suggestion](https://github.com/bitcoin/bitcoin/pull/31244#discussion_r2154811811) was tackled. Just left two code comment nits below, feel free to ignore.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/31244/commits/cf6e4170fa0529c4c238b7f185fc708e94dec34f

theStack reviewed (approved):
re-ACK cf6e4170fa0529c4c238b7f185fc708e94dec34f

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -1799,10 +1800,168 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+            }
+        } else if (paths.size() > 1) {
+            // All key provider vectors should be length 1. Clone them until they have the same length as paths
+            if (!clone_providers(paths.size())) {
+                error = "musig(): Multipath derivation path with multipath participants is disallowed"; // This error is unreachable due to earlier check
+                return {};
+            }
```
> // All key provider vectors should be length 1.

> // This error is unreachable due to earlier check

Instead of having uncovered code because it's unreachable, can't we replace the error setting with an assert instead?

```diff
- if (!clone_providers(paths.size())) {
-                error = "musig(): Multipath derivation path with multipath participants is disallowed"; // This error is unreachable due to earlier check
-                return {};
-            }
+ assert(max_providers_len == 1);
+ clone_providers(paths.size());
```

achow101 replied:
I've changed it to an `Assume`. I don't really like having `assert`s in modules like descriptors.

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -1799,10 +1800,168 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+            ret.emplace_back(std::make_unique<MuSigPubkeyProvider>(key_exp_index, std::move(pubs), path, deriv_type));
+        };
+
+        if (max_providers_len > 1 && paths.size() > 1) {
```
It took me some time to understand how the 4 conditions starting from here are handled individually. I feel some verbose and contextualised naming could be helpful for the reader as most of them are concerned with multipaths.

```
max_providers_len -> providers_max_multipaths
paths -> derivation_multipaths
```

This or something similar would be helpful imho.

achow101 replied:
Done, mostly

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -1799,10 +1800,168 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+
+        // Makes sure that all providers vectors in providers are the given length, or exactly length 1
+        // Length 1 vectors have the single provider cloned until it matches the given length.
+        const auto& clone_providers = [&providers](size_t length) -> bool {
+            for (auto& vec : providers) {
```
```
providers -> pubkeys_providers
vec -> pubkey_providers_mulltipath
```

Because `std::vector<std::vector<std::unique_ptr<PubkeyProvider>>>` is a vector of vectors representing multiple pubkeys in the Musig expression each of which could have multipaths if `vec.size() > 1`.
Applicable inside `emplace_final_provider` function as well.

I feel the `clone_providers` function is quite cool because IIUC it handles both the cases of expanding/cloning when either the pubkey provider multipath size is passed, or the musig derivation multipath size is passed.

achow101 replied:
Done, mostly

rkrux reviewed (commented):
Code review effe4fcc59bb2f84139bbaee064bfcb72da0baba

I have completed reviewing the `descriptor: Parse musig() key expressions` commit by going through it in detail and have suggested few naming suggestions if they make sense. 

I'm taking a final look at the `descriptor: Add MuSigPubkeyProvider` commit and the open discussions. Thank you for addressing the comments previously.

rkrux reviewed (commented):
I have finished reviewing the `descriptor: Add MuSigPubkeyProvider` commit - 4039be3f9b78c1ddb75c04f4a8002d36c4b05f79

achow101 committed:
descriptors: Move DeriveType parsing into its own function

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1799,10 +1800,168 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+            if (std::ranges::equal(deriv_split.back(), std::span{"*"}.first(1))) {
+                deriv_split.pop_back();
+                deriv_type = DeriveType::UNHARDENED;
+            } else if (std::ranges::equal(deriv_split.back(), std::span{"*'"}.first(2)) || std::ranges::equal(deriv_split.back(), std::span{"*h"}.first(2))) {
```
Ended up extracting this entire derive type parsing into a separate function.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1799,10 +1800,168 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+                error = strprintf("musig(): expected ',', got '%c'", expr[0]);
+                return {};
+            }
+            first = false;
```
Renamed

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1799,10 +1800,168 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+            if (IsKeyPathsHardened(paths)) {
+                error = "musig(): cannot have hardened derivation steps";
+                return {};
+            }
```
Added a `has_hardened` output parameter.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,218 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            pubout = *pub;
+            out.aggregate_pubkeys.emplace(m_aggregate_pubkey.value(), pubkeys);
+        } else {
+            if (!Assume(m_ranged_participants)) return std::nullopt;
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -580,6 +581,218 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+            CKeyID keyid = aggregate_pubkey->GetID();
+            std::copy(keyid.begin(), keyid.begin() + sizeof(info.fingerprint), info.fingerprint);
+            out.origins.emplace(keyid, std::make_pair(*aggregate_pubkey, info));
+            out.pubkeys.emplace(aggregate_pubkey->GetID(), *aggregate_pubkey);
```
Done

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/31244/commits/5fe7915c865a8e7c0a95ec376d0f1ff737f5d1c2

rkrux commented on src/pubkey.cpp:
```diff
@@ -197,20 +197,26 @@ constexpr XOnlyPubKey XOnlyPubKey::NUMS_H{
 };
 
-std::vector<CKeyID> XOnlyPubKey::GetKeyIDs() const
+std::vector<CPubKey> XOnlyPubKey::GetCPubKeys() const
```
Review note: Besides the internal use within `GetKeyIDs`, I see this function is actually used in the Musig signing PR.

rkrux reviewed (approved):
ACK 5fe7915c865a8e7c0a95ec376d0f1ff737f5d1c2

TYVM for patiently addressing all the comments in this big PR!

In the `tests: Test musig() parsing` commit, I have looked at all the unit test vectors but I have only glanced through the changes in the `DoCheck` function in the `descriptor_tests` for now.

theStack reviewed (approved):
re-ACK 5fe7915c865a8e7c0a95ec376d0f1ff737f5d1c2 🎹 

(as per `$ git range-diff cf6e417...5fe7915`, changes were mostly tackling rkrux' suggestions, which I agree are all improvements 👍 )
micro-nit: not worth it now to further invalidate acks, but fwiw, I'm still not a huge fan of cluttering the `ParseKeyPath{,num}` functions with extra booleans to restrict/detect properties that can be trivially detected after and think the previously suggested approach (`IsKeyPathsHardened` helper) was fine.

Sjors reviewed (commented):
Reviewed from 4af0dca096ca497a6b4e5314c9edea683efe620e to a53924bee321f9d01d053cf562ee3d9493e00529.

In a53924bee321f9d01d053cf562ee3d9493e00529 _descriptor: Parse musig() key expressions_: it might be slightly easier to follow if you extract `& has_hardened`, the `ParsePubkey` signature change and `FromString` changes into their own commit(s).

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -1802,9 +1811,154 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
+        bool any_ranged = false;
+        bool all_bip32 = true;
+        std::vector<std::vector<std::unique_ptr<PubkeyProvider>>> providers;
+        bool any_key_parsed = true;
```
Agree, see https://github.com/bitcoin/bitcoin/pull/31244#discussion_r2166251339

achow101 replied:
Will change if I need to retouch.

Sjors reviewed (approved):
ACK 5fe7915c865a8e7c0a95ec376d0f1ff737f5d1c2

I would suggest a followup PR to improve various code comments and such.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1647,14 +1649,15 @@ std::optional<uint32_t> ParseKeyPathNum(std::span<const char> elem, bool& apostr
  * @param[in] allow_multipath Allows the parsed path to use the multipath specifier
  * @returns false if parsing failed
```
>  it only ratchets hardened from false to true

No? It resets it to false first.

Sjors replied:
Ah yes, in that case:

```
@param[out] has_hardened whether hardened derivation was found
```

Sjors replied:
The version in https://github.com/bitcoin/bitcoin/pull/29675/commits/24bcdd27fd6c32646eb87eb8bea6f539bbc54e42 is fine.
